tls plugin seems to ignore config

### system info

Haraka | Haraka.js — Version: 2.8.15
 --- | :---
Node | v6.10.3
OS | Linux 7cc78a8c8d99 4.9.36-moby #1 SMP Wed Jul 12 15:29:07 UTC 2017 x86_64 Linux
openssl | OpenSSL 1.0.2k  26 Jan 2017

### Expected behavior
Given the following, I expect the server to start without errors:

```
/ # haraka -i /server
create: /server
create: /server/plugins
create: /server/docs
create: /server/config
create: /server/config/smtp.ini
create: /server/config/log.ini
create: /server/config/plugins
create: /server/config/dkim
create: /server/config/dkim/dkim_key_gen.sh
/ # mkdir -p /server/config/foo
/ # cd /server/config/foo
/server/config/foo # openssl req -x509 -nodes -days 2190 -newkey rsa:2048 -keyout cert.pem -out key.pem
Generating a 2048 bit RSA private key
....+++
...............+++
writing new private key to 'cert.pem'
-----
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) [AU]:.
State or Province Name (full name) [Some-State]:.
Locality Name (eg, city) []:
Organization Name (eg, company) [Internet Widgits Pty Ltd]:.
Organizational Unit Name (eg, section) []:
Common Name (e.g. server FQDN or YOUR name) []:example.com
Email Address []:
/server/config/foo # openssl dhparam -out dhparams.pem 1024
Generating DH parameters, 1024 bit long safe prime, generator 2
This is going to take a long time
...........................+..........................................................................................................................+...........................................................................................+........................+......................+............+.........+............................................................................+..............................................................................+..................................................................................................................................................................+.......................+...................................+..................+.......................+.............................+..............+..........................................................................................+.........................+......................................................................................................................................+........................................+................................................................+............+.........+........................+.................+..................+......+.............+...................+.................................................................................................................................................................................+..........................................+.....+...+.....................+......+...+...+.......................................................................................................................++*++*++*
/server/config/foo # ls -l
total 12
-rw-r--r--    1 root     root          1704 Sep 13 19:31 cert.pem
-rw-r--r--    1 root     root           245 Sep 13 19:34 dhparams.pem
-rw-r--r--    1 root     root          1099 Sep 13 19:31 key.pem
/server/config/foo # cd ..
/server/config # grep tls plugins
# see 'haraka -h tls' for config instructions before enabling!
tls
/server/config # cat tls.ini
key=foo/key.pem
cert=foo/cert.pem
dhparam=foo/dhparam.pem
```

### Observed behavior

However,

```
/server/config # haraka -c /server
Returning boolean false for main.timestamps=false
loaded TLD files: 1=1544 2=6156 3=2246
loaded 8357 Public Suffixes
Returning boolean true for main.reject=true
Returning boolean false for main.allow_mx_ip=0
Returning boolean true for main.reject_no_mx=1
loglevel: INFO
log format: DEFAULT
Starting up Haraka version 2.8.15
[INFO] [-] [core] Loading plugins
[INFO] [-] [core] Loading plugin: access
[INFO] [-] [access] skipping helo.checks.regexps
[INFO] [-] [core] Loading plugin: dnsbl
[INFO] [-] [core] Loading plugin: helo.checks
[INFO] [-] [core] Loading plugin: tls
[INFO] [-] [core] loading tls.ini
[ERROR] [-] [core] tls key tls_key.pem could not be loaded.
[ERROR] [-] [core] Config {
  root_path: '/server/config',
  module_config: [Function],
  overrides_path: '/server/config' }
[ERROR] [-] [core] tls cert tls_cert.pem could not be loaded.
[INFO] [-] [core] Loading plugin: mail_from.is_resolvable
[INFO] [-] [core] Loading plugin: rcpt_to.in_host_list
[INFO] [-] [core] Loading plugin: data.headers
[INFO] [-] [core] Loading plugin: queue/smtp_forward
[INFO] [-] [core] Loading plugin: max_unrecognized_commands
[NOTICE] [-] [core] Listening on :::25
[INFO] [-] [core] [outbound] Loading outbound queue from /server/queue
[WARN] [-] [dnsbl] zone 'zen.spamhaus.org' did not respond to test point (null)
[WARN] [-] [dnsbl] disabling zone 'zen.spamhaus.org'
```

--- a/Changes.md
+++ b/Changes.md
@@ -8,6 +8,7 @@
     * Replace console.log with stdout #2100
     * ES6: replace var with const or let  #2073
 * Fixes
+    * tls: only apply default key/cert paths when undefined #2111
     * haraka was adding TLS header on non-TLS connection #2103
     * dkim_verify: fix formatting of auth results #2107
     * smtp_forward: consistently use queue.wants #2107

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -330,6 +330,7 @@ exports.load_default_opts = function () {
     const cfg = certsByHost['*'];
 
     if (cfg.dhparam && typeof cfg.dhparam === 'string') {
+        log.logdebug(`loading dhparams from ${cfg.dhparam}`);
         tlss.saveOpt('*', 'dhparam', tlss.config.get(cfg.dhparam, 'binary'));
     }
 

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -293,7 +293,8 @@ exports.applySocketOpts = function (name) {
             tlss.saveOpt(name, opt, tlss.cfg[name][opt]);
             return;
         }
-        else if (tlss.cfg.main[opt] !== undefined) {
+
+        if (tlss.cfg.main[opt] !== undefined) {
             // if the setting exists in tls.ini [main]
             // then save it to the certsByHost options
             tlss.saveOpt(name, opt, tlss.cfg.main[opt]);

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -513,7 +513,7 @@ exports.ensureDhparams = function (done) {
 
     log.loginfo(`Generating a 2048 bit dhparams file at ${fpResolved}`);
 
-    const o = spawn('openssl', ['dhparam', '-out', `${fpResolved}`, '1024']);
+    const o = spawn('openssl', ['dhparam', '-out', `${fpResolved}`, '2048']);
     o.stdout.on('data', data => {
         // normally empty output
         log.logdebug(data);

--- a/tls_socket.js
+++ b/tls_socket.js
@@ -291,11 +291,13 @@ exports.applySocketOpts = function (name) {
         if (tlss.cfg[name] && tlss.cfg[name][opt] !== undefined) {
             // if the setting exists in tls.ini [name]
             tlss.saveOpt(name, opt, tlss.cfg[name][opt]);
+            return;
         }
         else if (tlss.cfg.main[opt] !== undefined) {
             // if the setting exists in tls.ini [main]
             // then save it to the certsByHost options
             tlss.saveOpt(name, opt, tlss.cfg.main[opt]);
+            return;
         }
 
         // defaults
@@ -506,11 +508,12 @@ exports.ensureDhparams = function (done) {
         return done(null, certsByHost['*'].dhparam);
     }
 
-    let filePath = tlss.cfg.main.dhparam;
-    if (!filePath) filePath = path.resolve(exports.config.root_path, 'dhparams.pem');
-    log.loginfo(`Generating a 2048 bit dhparams file at ${filePath}`);
+    const filePath = tlss.cfg.main.dhparam || 'dhparams.pem';
+    const fpResolved = path.resolve(exports.config.root_path, filePath);
 
-    const o = spawn('openssl', ['dhparam', '-out', `${filePath}`, '2048']);
+    log.loginfo(`Generating a 2048 bit dhparams file at ${fpResolved}`);
+
+    const o = spawn('openssl', ['dhparam', '-out', `${fpResolved}`, '1024']);
     o.stdout.on('data', data => {
         // normally empty output
         log.logdebug(data);
@@ -525,7 +528,7 @@ exports.ensureDhparams = function (done) {
             return done('Error code: ' + code);
         }
 
-        log.loginfo(`Saved to ${filePath}`);
+        log.loginfo(`Saved to ${fpResolved}`);
         const content = tlss.config.get(filePath, 'binary');
 
         tlss.saveOpt('*', 'dhparam', content);

